### PR TITLE
8330182: Start of release updates for JDK 24

### PR DESCRIPTION
**Pull request by jddarcy**
Get JDK 24 underway.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8330190](https://bugs.openjdk.org/browse/JDK-8330190) to be approved
- [x] Change requires CSR request [JDK-8330194](https://bugs.openjdk.org/browse/JDK-8330194) to be approved
- [x] Change requires CSR request [JDK-8330188](https://bugs.openjdk.org/browse/JDK-8330188) to be approved

### Issues
 * [JDK-8330182](https://bugs.openjdk.org/browse/JDK-8330182): Start of release updates for JDK 24 (**Bug** - P4)
 * [JDK-8330183](https://bugs.openjdk.org/browse/JDK-8330183): Add SourceVersion.RELEASE_24 (**Sub-task** - P4)
 * [JDK-8330184](https://bugs.openjdk.org/browse/JDK-8330184): Add source 24 and target 24 to javac (**Sub-task** - P4)
 * [JDK-8330190](https://bugs.openjdk.org/browse/JDK-8330190): Add SourceVersion.RELEASE_24 (**CSR**)
 * [JDK-8330194](https://bugs.openjdk.org/browse/JDK-8330194): Add source 24 and target 24 to javac (**CSR**)
 * [JDK-8330188](https://bugs.openjdk.org/browse/JDK-8330188): Start of release updates for JDK 24 (**CSR**)

### Reviewers
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [dcec9e96](https://git.openjdk.org/jdk/pull/18787/files/dcec9e964c212fe8c5ad73dcc9274de38beef54b)
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**) ⚠️ Review applies to [dc488f21](https://git.openjdk.org/jdk/pull/18787/files/dc488f21a76f9ebec40d44be88b48fb1cb88b49c)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [dc488f21](https://git.openjdk.org/jdk/pull/18787/files/dc488f21a76f9ebec40d44be88b48fb1cb88b49c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18787/head:pull/18787` \
`$ git checkout pull/18787`

Update a local copy of the PR: \
`$ git checkout pull/18787` \
`$ git pull https://git.openjdk.org/jdk.git pull/18787/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18787`

View PR using the GUI difftool: \
`$ git pr show -t 18787`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18787.diff">https://git.openjdk.org/jdk/pull/18787.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18787#issuecomment-2057616108)

### CODE REVIEW

**Comment by jddarcy** [comment]
This initial version of the PR intentionally excludes the creation of the new symbol files so that the fundamental code aspects of the update are easier to see.

File: make/conf/version-numbers.conf [hunk 1]
```diff
@@ -26,17 +26,17 @@
...
+DEFAULT_VERSION_FEATURE=24
 DEFAULT_VERSION_INTERIM=0
 DEFAULT_VERSION_UPDATE=0
 DEFAULT_VERSION_PATCH=0
 DEFAULT_VERSION_EXTRA1=0
 DEFAULT_VERSION_EXTRA2=0
 DEFAULT_VERSION_EXTRA3=0
-DEFAULT_VERSION_DATE=2024-09-17
-DEFAULT_VERSION_CLASSFILE_MAJOR=67  # "`$EXPR $DEFAULT_VERSION_FEATURE + 44`"
+DEFAULT_VERSION_DATE=2024-03-17
```
Comment (irisclark) [inline make/conf/version-numbers.conf:36]: "2024-03-17"?  I thought that the expected date was "2025-03-18"?

**Review by irisclark — APPROVED** [review]
The copyright year was  not updated in all files *14.java.  I assume that's intentional. 

I've also Reviewed the associated CSRs.

File: make/conf/version-numbers.conf (hunk 1 above)
Comment (jddarcy) [reply make/conf/version-numbers.conf:36]: Thanks for catching that; yes, I'll correct and double-check the expected GA date in the next push.

**Comment by jddarcy** [comment]
> The copyright year was not updated in all files *14.java. I assume that's intentional.

I'll run my copyright update script before pushing.

> I've also Reviewed the associated CSRs.

Thanks.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

File: src/java.base/share/classes/java/lang/classfile/ClassFile.java [hunk 2]
```diff
@@ -1478,6 +1478,9 @@ default List<VerifyError> verify(Path path) throws IOException {
     /** 67 */
     int JAVA_23_VERSION = 67;
 
+    /** 68 */
```
Comment (liach) [inline src/java.base/share/classes/java/lang/classfile/ClassFile.java:1481]: We need `@since 24` here.

Comment (jddarcy) [reply src/java.base/share/classes/java/lang/classfile/ClassFile.java:1481]: Ah, good catch; looks like I was treating that as an internal API element when it is indeed a public API element. I've filed JDK-8330458 to fix-up the Java 23 constant as well.

**Review by asotona — APPROVED** [review]
ClassFile changes are OK.

**Review by irisclark — APPROVED** [review]

File: test/hotspot/jtreg/runtime/CommandLine/VMDeprecatedOptions.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/CommandLine/VMDeprecatedOptions.java:1]: There are further updates to this test in the pipeline (new deprecated flags in 23) so you will need to keep updating to reflect that.

**Review by dholmes-ora — APPROVED** [review]
LGTM

Thanks

File: test/hotspot/jtreg/runtime/CommandLine/VMDeprecatedOptions.java
Comment (jddarcy) [reply test/hotspot/jtreg/runtime/CommandLine/VMDeprecatedOptions.java:1]: > There are further updates to this test in the pipeline (new deprecated flags in 23) so you will need to keep updating to reflect that.

Thanks @dholmes-ora ; acknowledged.

**Review by irisclark — APPROVED** [review]

**Review by irisclark — APPROVED** [review]
Still looks good.

**Review by irisclark — APPROVED** [review]

File: src/jdk.compiler/share/data/symbols/jdk.incubator.foreign-N.sym.txt
Comment (liach) [inline src/jdk.compiler/share/data/symbols/jdk.incubator.foreign-N.sym.txt:1]: Just curious, does CreateSymbols not track module migrations, now that jdk.incubator.foreign is completely merged into java.base?

Comment (lahodaj) [reply src/jdk.compiler/share/data/symbols/jdk.incubator.foreign-N.sym.txt:1]: When writing these symbol files, CreateSymbols does not keep track where a given package was in a given version, and puts packages to files based on the first version where the package was first introduced. Technically, it should not matter, as the assignment of classes/packages to modules is not based on the file from which the description of the class was read. So, so far, it didn't seem necessary to keep track of package movement for writing of these files. But it is feasible to do so, if the need arises. (When writing ct.sym, it keeps track of the exact package module for every version, of course.)

**Review by irisclark — APPROVED** [review]

**Pull request closed without merging**